src/installed_dune_file.ml

```diff
@@ -21,33 +21,54 @@ let parse_sub_systems sexps =
   |> Sub_system_name.Map.mapi ~f:(fun name (_, version, data) ->
     let (module M) = Jbuild.Sub_system_info.get name in
     Syntax.check_supported M.syntax version;
-    let parsing_context =
-      Univ_map.singleton (Syntax.key M.syntax)
-        (* This is wrong, see #909 *)
-        (0, 0)
-    in
     M.T (Sexp.Of_sexp.parse M.parse parsing_context data))
 
 let of_sexp =
   let open Sexp.Of_sexp in
   let version =
     plain_string (fun ~loc -> function
-      | "1" -> ()
-      | _ ->
+      | "1" | "2" -> ()
+      | v ->
         of_sexp_errorf loc
-          "Unsupported version, only version 1 is supported")
+          "Unsupported version %S, only version 1 is supported" v)
   in
   sum
     [ "dune",
       (version >>= fun () ->
-       list raw >>| fun l ->
-       parse_sub_systems l)
+       get_all >>= fun parsing_context ->
+       list raw >>|
+       parse_sub_systems ~parsing_context)
     ]
 
 let load fname =
-  Sexp.Of_sexp.parse of_sexp Univ_map.empty (Io.Sexp.load ~mode:Single fname)
+  Io.with_lexbuf_from_file fname ~f:(fun lexbuf ->
+    let (version_loc, version) =
+      let bad_dune_file = "Unable to read (dune x.y ..) line file" in
+      let rec loop = function
+        | [_; _; _] as a -> List.rev a
+        | acc ->
+          begin match (Sexp.Lexer.token lexbuf : Sexp.Lexer.Token.t) with
+          | Eof ->
+            Loc.fail (Loc.in_file (Path.to_string fname)) "%s" bad_dune_file
+          | t -> loop (t :: acc)
+          end
+      in
+      let loc = Sexp.Loc.of_lexbuf lexbuf in
+      match loop [] with
+      | [Lparen; Atom (A "dune"); Atom s] -> (loc, Sexp.Atom.to_string s)
+      | _ -> Loc.fail loc "%s" bad_dune_file
+    in
+    let (lexer, syntax) =
+      match version with
+      | "1" -> (Sexp.Lexer.jbuild_token, (0, 0))
+      | "2" -> (Sexp.Lexer.token, (1, 0))
+      | _   -> Loc.fail version_loc "unknown version %S" version
+    in
+    Sexp.Of_sexp.parse of_sexp
+      (Univ_map.singleton (Syntax.key Stanza.syntax) syntax)
+      (Io.Sexp.load ~lexer ~mode:Single fname))
```

Comment:
We shouldn't reopen the file a second time here. We need to construct a `lexer` function that first return the three extracted token and then switch to the real lexer.

Reply:
I'm not really sure how to do it the way you say however. Seems like we'd need to call parser with an initial set of lexemes (the 3 lexemes that we already parsed).

What should be possible is just reading the file into a string and then constructing the lexer from_string a second time. That should cut down on the sys calls.

Reply:
There are two ways of doing it:

first one:

```ocaml
let pending_tokens = ref ... in
let lexer lb =
  match !pending_tokens with
  | [] -> lexer lb
  | x :: l -> pending_tokens := l; x
in
...
```

second one, which implement all the logic in one function:

```ocaml
let state = ref 0 in
let lexer = ref !Sexp.Lexer.token in
let lexer lb =
  let token = !lexer lb in
  (match !state, token with
   | 0, Lparen -> state := 1
   | 1, Atom (A "dune") -> state := 2
   | 2, Atom (A "1") -> state := 3; lexer := Sexp.Lexer.jbuild_token
   | 2, Atom (A "2") -> state := 3;
   | 2, Atom (A s) -> error ()
   | 3, _ -> ()
   | _ -> error ());
  token
in
```

Reply:
I see. I went for the first approach. Did not it was allowed to reuse the same state between possibly different lexer functions like this.